### ISSUE TITLE
--cores=all not working correctly : binaries relying on openmp still single-threaded

### ISSUE THREAD

**Issue by jtressle** [snakemake/snakemake#521]
I'm testing on snakemake version 5.20.1.  

I have a pipeline that uses multiple c++ programs compiled with openmp in a docker container.  These binaries are compiled in cmake 3.9 using the following lines to enable openMP:

find_package(OpenMP)
if(OpenMP_CXX_FOUND)
    target_link_libraries (binary_name PUBLIC OpenMP::OpenMP_CXX)
endif()

In snakemake version 5.10 and lower, I run my snakemake script without specifying any cores.  I verified I am using all cores as htop is saying docker is using 400% CPU.

When I run the same script in 5.20.1, my CPU usage doesn't go above 120% for my binaries.  I get the same result if I use '--cores=all', '--cores=4', '-j all', and '-j 4'.

Unfortunately I am unable to share my binaries, but I can see about creating a c++ test binary if needed.  

My workaround now is to use snakemake version 5.10.

Thanks

**Comment by johanneskoester** [comment]
At some point between those releases, I think we made a change that snakemake automatically restricts the paralelelism of openmp based on the threads property of the rule. So, you will need to set `threads: n` where n is the desired amount of paralelism that you want.

**Comment by johanneskoester** [comment]
Sorry, this did not make it into the docs. @dlaehnemann can you add a section explaining what you do for constraining [here](https://snakemake.readthedocs.io/en/stable/snakefiles/rules.html#threads)?

**Comment by jtressle** [comment]
Thanks for the update.

So what is the default behavior?  Shouldn't it be if I define cores=all, threads should be maximized unless specified otherwise?

**Comment by dlaehnemann** [comment]
I think the misunderstanding might be about where you specify what:

1. You specify `--cores all` in the snakemake command (`snakemake --cores all`) to globally limit the amount of cores that snakemake can use (capped at the actual amount of threads available).
2. You specify `threads: 8` to limit a rule's execution to 8 threads (or actually, to declare that a rule uses up to 8 threads and snakemake will "reserve" those for you as part of the total limit that you provided it with). While snakemake is usually permissive in letting rules use any number of threads if no `threads:` argument is given, we have recently limited this in the case of stuff parallelized by openmp or linear algebra libraries. These were simply grabbing all cores available on a system for each execution of a rule and once you have parallel executions of such a rule, you the pipeline starts using the same set of available cores multiple times. Eventually, this leads to chaotic executions that can start blocking each other. To make this controllable, the default is now that such libraries are limited to the number of cores provided to the respective rule, with a default of 1 core if no `thread: ` specification is provided.

Long story short: Keep setting `snakemake --cores all` to  provide all available cores to your pipeline. Set `threads: XYZ` to give the rule that should parallelize internally via openmp XYZ cores to use.

If you're interested in the background, check out PR #249 .
An addition to the docs is incoming...

**Comment by jtressle** [comment]
Understood.  Unfortunately, for my scenario, my rules all run openMP binaries so each rule will need this threading entry.

What will be the correct threading value I need to replicate 5.10?
Is it `threads: workflow.cores * 1.0`?

Much thanks for the explanation.

**Comment by dlaehnemann** [comment]
The `threads_max` is simply an example of some variable that you precompute yourself, whereas `workflow.cores` is an attribute of the workflow that is always available automatically. I think `threads: workflow.cores` should do the trick.

Did you not run into slow-downs with the previous behaviour, because parallel executions of your rules were grabbing the same cores multiple times and overloading them? Simply curious, here...

And sorry for the inconvenience caused, but this was previously uncontrollable behaviour...

**Comment by dlaehnemann** [comment]
Ah, and feel free to close this issue, if this answers your question. Or simply keep on asking! :D

**Comment by jtressle** [comment]
Thanks for your help.  Closing the issue.  

Actually, in my implementation, I'm use snakemake in it's simplest form and run sequential pipelines.  So any parallelism occurs inside my binaries.  I haven't had the need to run parallel executions of my rules yet.

**Comment by dlaehnemann** [comment]
Ah, OK. I can't imagine living without that functionality any more! :D

I'm glad this solves it for you.
